libs/utils/src/tabulation.ts

```diff
@@ -0,0 +1,336 @@
+  election: Election
+): OfficialCandidateNameLookup {
+  const lookupInternal: Record<ContestId, Record<CandidateId, string>> = {};
+  for (const contest of election.contests) {
+    if (contest.type === 'candidate') {
+      const contestCandidateLookup: Record<CandidateId, string> = {};
+      for (const candidate of contest.candidates) {
+        contestCandidateLookup[candidate.id] = candidate.name;
+      }
+      lookupInternal[contest.id] = contestCandidateLookup;
+    }
+  }
+
+  function get(contestId: ContestId, candidateId: CandidateId): string {
+    return assertDefined(assertDefined(lookupInternal[contestId])[candidateId]);
+  }
+
+  return {
+    get,
+  };
+}
+
+function getCastVoteRecordGroupSpecifier(
+  cvr: Tabulation.CastVoteRecord,
+  groupBy: Tabulation.GroupBy,
+  partyIdLookup: BallotStyleIdPartyIdLookup
+): Tabulation.GroupSpecifier {
+  return {
+    ballotStyleId: groupBy.groupByBallotStyle ? cvr.ballotStyleId : undefined,
+    precinctId: groupBy.groupByPrecinct ? cvr.precinctId : undefined,
+    batchId: groupBy.groupByBatch ? cvr.batchId : undefined,
+    scannerId: groupBy.groupByScanner ? cvr.scannerId : undefined,
+    votingMethod: groupBy.groupByVotingMethod ? cvr.votingMethod : undefined,
+    partyId: groupBy.groupByParty
+      ? partyIdLookup[cvr.ballotStyleId]
+      : undefined,
+  };
+}
+
+const GROUP_KEY_SEPARATOR = '&';
+export const GROUP_KEY_ROOT: GroupKey = 'root';
+
+function escapeGroupKeyPart(groupKeyPart: string): string {
+  return groupKeyPart.replaceAll('\\', '\\\\').replaceAll('&', '\\&');
+}
+
+/**
+ * Based on a group's attributes, defines a key which is used to
+ * lookup and identify grouped election results.
+ */
+export function getGroupKey(
+  groupSpecifier: Tabulation.GroupSpecifier,
+  groupBy: Tabulation.GroupBy
+): Tabulation.GroupKey {
+  const keyParts: string[] = [GROUP_KEY_ROOT];
+  if (groupBy.groupByBallotStyle) {
+    keyParts.push(assertDefined(groupSpecifier.ballotStyleId));
+  }
+
+  if (groupBy.groupByParty) {
+    keyParts.push(assertDefined(groupSpecifier.partyId));
+  }
+
+  if (groupBy.groupByBatch) {
+    keyParts.push(assertDefined(groupSpecifier.batchId));
+  }
+
+  if (groupBy.groupByScanner) {
+    keyParts.push(assertDefined(groupSpecifier.scannerId));
+  }
+
+  if (groupBy.groupByPrecinct) {
+    keyParts.push(assertDefined(groupSpecifier.precinctId));
+  }
+
+  if (groupBy.groupByVotingMethod) {
+    keyParts.push(assertDefined(groupSpecifier.votingMethod));
+  }
+
+  return keyParts.map(escapeGroupKeyPart).join(GROUP_KEY_SEPARATOR);
+}
+
+/**
+ * Tabulates iterable cast vote records into election results, grouped by
+ * the attributes specified {@link Tabulation.GroupBy} parameter.
+ */
+export function tabulateCastVoteRecords({
+  election,
+  cvrs,
+  groupBy,
+}: {
+  cvrs: Iterable<Tabulation.CastVoteRecord>;
+  election: Election;
+  groupBy?: Tabulation.GroupBy;
+}): Tabulation.GroupedElectionResults {
```

Comment:
One requirement of our tabulation stack is that we're able to efficiently tabulate grouped results. For example, all results grouped by batches (of which there could be hundreds or thousands). I think it's reasonable to build for supporting this in one database pass rather than one for each group.

As such, these methods operate in terms of GroupedElectionResults. I didn't want to fiddle with a true multi-key map which, under the hood, may be complicated. I used simply a flat dictionary with the keys being GroupKey strings which are just strings produced from a group specifier. Each ElectionResults also has its group metadata as a property (e.g. a ballot style id when grouping by ballot style), so the dictionary structure is really a lookup optimization rather than a necessary organization of the data.

Some thoughts:
- I originally had separate methods for tabulation without grouping, but I collapsed those into a special case of the principle methods which only return grouped results. Getting a dictionary with one key-value pair is annoying, but I think having these methods (of aggregating CVRs or write-in data) in only one place is better.
- The grouping is not very type-safe. If you group by ballot style, you don't get types that have a required ballot style property. You get a type which has an optional ballot style property and have to know from context what is required. A different approach would be to type this more strongly, but in thinking it through it sounded overly complex in terms of defining types and overloads in many places, so I opted not to do that.
- The grouping pattern is very flexible. You can group by one attribute or any combination of attributes. We could instead have a limited set of allowed groupings, but I don't think we know what those are in the product and we already know that we have composite groupings (e.g. every group + party for primaries, precinct + voting method for existing CSV exports) so I went the flexible route.